ibis/backends/dask/execution/util.py

```diff
@@ -355,6 +396,52 @@ def helper(
     return df
 
 
+def add_sequential_partitioned_sorted_column(
+    df: dd.DataFrame | dd.Series,
+) -> dd.DataFrame:
+    """Add a column that is already partitioned and sorted.
+
+    This column acts as if we had a global index across the distributed data.
+    This index needs to be consecutive in the range of [0, len(df)) allows
+    downstream operations to work properly.
+    The default index of dask dataframes is to be consecutive within each partition.
+
+    Important properties:
+
+    - Each row has a unique id (i.e. a value in this column)
+    - The global index that's added is sequential in the same order that the rows currently are in.
+    - IDs within each partition are already sorted
+
+    We also do not explicity deal with overflow in the bounds.
+
+    Parameters
+    ----------
+    df : dd.DataFrame
+        Dataframe to add the column to
+
+    Returns
+    -------
+    dd.DataFrame
+        New dask dataframe with sorted partitioned index
+    """
+    if isinstance(df, dd.Series):
+        df = df.to_frame()
+
+    col_name = "_ibis_index"
+
+    if col_name in df.columns:
+        raise ValueError(f"Column {col_name} is already present in DataFrame")
+
+    df = df.reset_index(drop=True)
+    df = df.assign(**{col_name: 1})
+    df = df.assign(**{col_name: df[col_name].cumsum() - 1})
+    df = df.set_index(
+        col_name,
+        sorted=True,
```

Comment:
If you don't pass a value for `divisions` I believe this triggers dask to compute the whole dataframe up to this point, which we very much want to avoid. I think we'll have to think of something different here.